File: ios/Shelf/Models/Series.swift
import Foundation

// MARK: - Series

struct Series: Codable, Identifiable, Hashable {
    let id: UUID
    let name: String
    let description: String?
    let totalBooks: Int?
    let isComplete: Bool
    let coverImageURL: String?
    let works: [SeriesWorkItem]

    enum CodingKeys: String, CodingKey {
        case id, name, description, works
        case totalBooks = "total_books"
        case isComplete = "is_complete"
        case coverImageURL = "cover_image_url"
    }
}

// MARK: - Series Work Item

struct SeriesWorkItem: Codable, Identifiable, Hashable {
    var id: UUID { workID }
    let position: Double
    let isMainEntry: Bool
    let workID: UUID
    let title: String
    let authors: [String]
    let coverImageURL: String?
    let userStatus: String?

    enum CodingKeys: String, CodingKey {
        case position, title, authors
        case isMainEntry = "is_main_entry"
        case workID = "work_id"
        case coverImageURL = "cover_image_url"
        case userStatus = "user_status"
    }
}

// MARK: - Series Progress

struct SeriesProgress: Codable {
    let seriesID: UUID
    let seriesName: String
    let totalMainEntries: Int
    let readCount: Int
    let readingCount: Int
    let progressPercent: Double

    enum CodingKeys: String, CodingKey {
        case seriesID = "series_id"
        case seriesName = "series_name"
        case totalMainEntries = "total_main_entries"
        case readCount = "read_count"
        case readingCount = "reading_count"
        case progressPercent = "progress_percent"
    }
}

File: backend/api/schemas/series.py
"""Schemas for book series endpoints."""

from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel


class SeriesWorkItem(BaseModel):
    position: float
    is_main_entry: bool
    work_id: UUID
    title: str
    authors: list[str]
    cover_image_url: str | None = None
    user_status: str | None = None

    model_config = {"from_attributes": True}


class SeriesResponse(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    total_books: int | None = None
    is_complete: bool
    cover_image_url: str | None = None
    works: list[SeriesWorkItem]

    model_config = {"from_attributes": True}


class SeriesProgressResponse(BaseModel):
    series_id: UUID
    series_name: str
    total_main_entries: int
    read_count: int
    reading_count: int
    progress_percent: float
